## Gather

In [1]:
import pandas as pd

In [2]:
patients = pd.read_csv('patients.csv')
treatments = pd.read_csv('treatments.csv')
adverse_reactions = pd.read_csv('adverse_reactions.csv')

## Assess

In [6]:
patients.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   
3         Woodbridge          NJ    7095.0  United States   
4             Dothan          AL   36303.0  United States   

                                      contact  birthdate  weight  height   bmi  
0        951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66  19.6  
1       PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66  19.2  
2            402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71  24.8  
3  PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   220.9      70  31.7  
4             334-515-7487TimNeudorf@cuvox.de  2/18/1928   192.3      27  26.1

In [7]:
treatments.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0           NaN  
1          0.97  
2           NaN  
3          0.35  
4          0.32

In [8]:
adverse_reactions.head()

given_name     surname           adverse_reaction
0      berta  napolitani  injection site discomfort
1       lena        baer               hypoglycemia
2     joseph         day               hypoglycemia
3     flavia  fiorentino                      cough
4    manouck     wubbels          throat irritation

In [9]:
patients.surname.value_counts()

Doe                  6
Jakobsen             3
Taylor               3
Lund                 2
Johnson              2
Lương                2
Souza                2
Gersten              2
Ogochukwu            2
Parker               2
Batukayev            2
Nilsen               2
Cabrera              2
Liễu                 2
Bùi                  2
Cindrić              2
Silva                2
Collins              2
Lâm                  2
Tạ                   2
Woźniak              2
Tucker               2
Aranda               2
Berg                 2
Dratchev             2
Hueber               2
Kadyrov              2
Schiavone            2
Kowalczyk            2
Grímsdóttir          2
                    ..
Amanuel              1
Eskelinen            1
Endrizzi             1
Nebay                1
Blix                 1
Shuler               1
Vesecká              1
Pérez                1
Fujiwara             1
Gibson               1
Vestergaard          1
Ekwueme              1
Hermansen  

In [10]:
patients.address.value_counts()

123 Main Street               6
648 Old Dear Lane             2
2778 North Avenue             2
2476 Fulton Street            2
1742 Tea Berry Lane           1
1812 Poplar Street            1
1079 Ingram Street            1
2397 Bell Street              1
4792 Maud Street              1
4704 Edsel Road               1
1865 Honeysuckle Lane         1
4605 Hall Street              1
3710 Jerry Dove Drive         1
212 Tibbs Avenue              1
3868 Freed Drive              1
1536 John Daniel Drive        1
2074 Parrish Avenue           1
3084 Blue Spruce Lane         1
3893 Eva Pearl Street         1
3539 Bottom Lane              1
959 Heritage Road             1
283 Simons Hollow Road        1
4291 Patton Lane              1
2103 Edington Drive           1
2884 Elsie Drive              1
3820 Terry Lane               1
3234 Sardis Station           1
4178 Despard Street           1
3073 Bedford Street           1
1589 Roosevelt Wilson Lane    1
                             ..
1813 Lin

In [11]:
patients[patients.address.duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

In [12]:
patients.weight.sort_values()

210     48.8
459    102.1
335    102.7
74     103.2
317    106.0
171    106.5
51     107.1
270    108.1
198    108.5
48     109.1
478    109.6
141    110.2
38     111.8
438    112.0
14     112.0
235    112.2
307    112.4
191    112.6
408    113.1
49     113.3
326    114.0
338    114.1
253    117.0
321    118.4
168    118.8
1      118.8
350    119.0
207    119.2
265    120.0
341    120.3
       ...  
332    224.0
252    224.2
12     224.2
222    224.8
166    225.3
111    225.9
101    226.2
150    226.6
352    227.7
428    227.7
88     227.7
13     228.4
339    229.0
182    230.3
121    230.8
257    231.7
395    231.9
246    232.1
219    237.8
11     238.7
50     238.9
441    239.1
499    239.6
439    242.0
487    242.4
144    244.9
61     244.9
283    245.5
118    254.5
485    255.9
Name: weight, Length: 503, dtype: float64

In [15]:
weight_lbs = patients[patients.surname == 'Zaitseva'].weight * 2.20462
height_in = patients[patients.surname == 'Zaitseva'].height
bmi_check = 703 * weight_lbs/(height_in * height_in)
bmi_check

210    19.055827
dtype: float64

In [17]:
patients[patients.surname == 'Zaitseva'].bmi

210    19.1
Name: bmi, dtype: float64

Observe the BMI of both are same so the 1st result of __patients.weight.sort_values()__ show the weight 48.8 is in kg instead of lbs. 

In [13]:
sum(treatments.auralin.isnull())

0

In [14]:
sum(treatments.novodra.isnull())

0

## Observe
- 'Doe' in surname occurs 6 times. Also __123 Main Street__ also occurs 6 times. Also observe the o/p of __patients[patients.address.duplicated()]__. __John Doe__ occurs number of times so its a __VALIDITY__ issue.  So add this issue to the list below.
- __patients[patients.address.duplicated()]__ reveals there are same address for a person where in one of his record his nick name was used : Jake Jacobsone and Jakob Jacobsone.
- patients weight in in kg (48.8) : `patients[patients.surname == 'Zaitseva'].weight`
- The programmatic assessment shows that there are no null datasets in __treatments.auralin__ and __treatments.novodra__. Bur there are lot of missing values in the columns but not considered as NULL by Pandas as they are represented using - , / or NA or text NULL. We cant use these for calculations. So null must be represented correctly. Again it is a __VALIDITY ISSUE__.

## Assess

## Quality
#### patients table
- zipcodes is a float not a string.
- zipcodes has 4 digits sometimes.
- Tim Neudorf height is 27 instead of 72 inches.
- full state names sometimes, abbreviations other times. Table must have consistent data.
- Date of birth as month/day/year instead of day/month/year isn't a data quality issue because either representation is fine and the entire birthdate column consists of one format.
- Dsvid Gustafsson
- __ Weight as a float type (one decimal) and height as an integer (no decimals) is fine. This is just how the data were recorded. Though height may therefore be technically slightly inaccurate (by portions of an inch), this is an example of an acceptable limitation.
- missing demographic info (address, contact,.. columns)
- Erroneous datatypes(sex, state, zipcode,birthdate)
- multiple phone number formats 
- Default John Doe data (Like a garbage data)
- Multiple records for Jakobsen, Gersten, Taylor
- kgs instead of lbs for Zaitseva weight


#### treatments table
- missing HbA1c changes.
- (Accents on given names and surnames (e.g., Tám Liễu) aren't data quality issues because this is how these names are actually spelled. You will have to be aware of these accents whenever performing operations that require this text (merging tables on the name column, for example), but they aren't quality issues.)
- letter u in the strating and ending doses for Auralin and Novodra.
- lowercase given names and surnames
- missing records (280 instead of 350)
- Erroneous datatypes(auralin and novadra).
- Inaccurate HbA1c changes (4s mistaken as 9s)
- Nulls represented as dashes (-) in auralin and novodra columns.

#### adverse_reactions table
- lower case given names and surnames

## Tidiness 

#### Done using visual assessment

- contact column in __patients table__ should be split into phone number and email.
- three variables in two columns in treatments table (treatment, start_dose and end_dose.)

# How Data Gets Dirty and Messy
## Sources of Dirty Data
_Dirty data = low quality data = content issues_

There are lots of sources of dirty data. Basically, anytime humans are involved, there's going to be dirty data. There are lots of ways in which we touch data we work with.

- We're going to have user entry errors.
- In some situations, we won't have any data coding standards, or where we do have standards they'll be poorly applied, causing problems in the resulting data
- We might have to integrate data where different schemas have been used for the same type of item.
- We'll have legacy data systems, where data wasn't coded when disc and memory constraints were much more restrictive than they are now. Over time systems evolve. Needs change, and data changes.
- Some of our data won't have the unique identifiers it should.
- Other data will be lost in transformation from one format to another.
- And then, of course, there's always programmer error.
- And finally, data might have been corrupted in transmission or storage by cosmic rays or other physical phenomenon. So hey, one that's not our fault.

## Sources of Messy Data
_Messy data = untidy data = structural issues_

Messy data is usually the result of poor data planning. Or a lack of awareness of the benefits of [tidy data](https://cran.r-project.org/web/packages/tidyr/vignettes/tidy-data.html). Fortunately, messy data is usually much more easily addressable than most of the sources of dirty data mentioned above.

# You Can Iterate

The concept of iterating isn't that applicable for clinical trials given the rigor involved in their planning. But, theoretically, the following situations could arise that require iteration:

- Maybe you (as the data analyst or data scientist on the clinical trial research team) realized your statistical power calculations were wrong, and you needed to recruit more patients to make your study statistically significant. You'd also have to do revisit gathering in this scenario.
- Maybe you realized you were missing a key piece of patient information, like patient blood type (again, unlikely given the rigor of clinical trials, but mistakes happen) because you discovered new research that related insulin resistance to blood type. You'd also have to do revisit gathering in this scenario.
- Maybe you finished assessing, started cleaning, and spotted another data quality issue. Revisiting assessing to add these assessments to your notes is fine.